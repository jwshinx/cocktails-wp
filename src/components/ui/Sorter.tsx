import _ from 'lodash'
import React from 'react'

import { PropertyType } from '../../models/cocktail'

function Sorter<T>({
  objects,
  onSortPropertyClick,
}: {
  objects: Array<T>
  onSortPropertyClick: (propertyType: PropertyType<T>) => void
}): JSX.Element {
  if (objects === undefined || objects === null || _.isEmpty(objects)) {
    return <></>
  }

  const object = objects.length > 0 ? objects[0] : {}

  const onSelectChange = (object: string): void => {
    // console.log("+++> Sorter onSelectChange fired!");
    const values: Array<string> = object.split('-')
    if (values.length === 2) {
      onSortPropertyClick({
        property: values[0] as keyof T,
        isDescending: values[1] === 'true',
      })
    }
  }

  return (
    <div className="form-group">
      <select
        className="form-select mb-3"
        id="sorter"
        onChange={(event) => onSelectChange(event.target.value)}
      >
        {Object.keys(object).map((key): JSX.Element => {
          return (
            <React.Fragment key={key}>
              <option key={`${key}-true`} value={`${key}-true`}>
                Sort by {key} descending
              </option>
              <option key={`${key}-false`} value={`${key}-false`}>
                Sort by {key} ascending
              </option>
            </React.Fragment>
          )
        })}
        ;
      </select>
    </div>
  )
}

export default Sorter
