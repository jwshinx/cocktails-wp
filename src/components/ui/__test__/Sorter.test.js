/**
 * @jest-environment jsdom
 */

import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'

import Sorter from '../Sorter'

const mockFn = jest.fn(() => true)

const data = [
  { idDrink: '1', strDrink: 'Martini' },
  { idDrink: '2', strDrink: 'Greyhound' },
  { idDrink: '3', strDrink: 'Mai Tai' },
  { idDrink: '4', strDrink: 'Rob Roy' },
]

it('should render a combobox  ', () => {
  render(<Sorter objects={data} onSortPropertyClick={mockFn} />)
  const combobox = screen.getByRole('combobox')
  expect(combobox).toBeInTheDocument()
})

it('should render 4 options: "idDrink" and "strDrink" -- "asc", "desc"', () => {
  render(<Sorter objects={data} onSortPropertyClick={mockFn} />)
  const options = screen.getAllByRole('option')
  expect(options.length).toBe(4)
})

it('should snapshot Sorter', () => {
  const tree = renderer
    .create(<Sorter objects={data} onSortPropertyClick={mockFn} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
