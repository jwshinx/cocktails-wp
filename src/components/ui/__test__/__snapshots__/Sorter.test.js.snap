// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should snapshot Sorter 1`] = `
<div
  className="form-group"
>
  <select
    className="form-select mb-3"
    id="sorter"
    onChange={[Function]}
  >
    <option
      value="idDrink-true"
    >
      Sort by 
      idDrink
       descending
    </option>
    <option
      value="idDrink-false"
    >
      Sort by 
      idDrink
       ascending
    </option>
    <option
      value="strDrink-true"
    >
      Sort by 
      strDrink
       descending
    </option>
    <option
      value="strDrink-false"
    >
      Sort by 
      strDrink
       ascending
    </option>
    ;
  </select>
</div>
`;
